### PR TITLE
Fix Hash#rehash

### PR DESCRIPTION
**Pull request by asterite**
It wasn't working properly: entries were rehashed but an old hash value was used to compute the index where they should go.

### CODE REVIEW

File: src/hash.cr [hunk 1]
```diff
@@ -572,9 +572,12 @@ class Hash(K, V)
     # while moving non-deleted entries to the beginning (compaction).
     new_entry_index = 0
     each_entry_with_index do |entry, entry_index|
+      entry_hash = entry.hash
```
Comment (straight-shoota) [inline src/hash.cr]: Wouldn't this be better placed in the `else` branch?

Comment (asterite) [reply src/hash.cr]: Maybe, I'm not sure what's the difference.

Comment (Sija) [reply src/hash.cr]: In this way you'd avoid one unnecessary call.
I guess that was @straight-shoota's point.

Comment (asterite) [reply src/hash.cr]: Sounds good!

(just note that `entry.hash` is just accessing an instance variable from a struct and that's like super cheap, plus I'm almost sure that LLVM will figure out that `entry_hash` is only reassigned in the `if` branch and so the right thing)

Comment (straight-shoota) [reply src/hash.cr]: Performance shouldn't be an issue. But it's easier to understand the code. This assignment is only relevant in the else branch. So it should only be there.

**Pull request merged**